service/src/main/java/com/ge/predix/acs/encryption/Encryptor.java

```diff
@@ -20,31 +22,69 @@
     private static final String ALGO_WITH_PADDING = "AES/CBC/PKCS5PADDING";
     private static final String ALGO = "AES";
-    private static final String ENCODING = "UTF-8";
     private static final int IV_LENGTH_IN_BYTES = 16;
     private static final int KEY_LENGTH_IN_BYTES = 16;
+    private static final ThreadLocal<CachedCipher> CIPHER_THREAD_LOCAL = new ThreadLocal<>();
 
-    private Cipher cipher;
-    private String encryptionKey;
+    private SecretKeySpec secretKeySpec;
 
-    public Encryptor() {
+    // Modified version of Cassandra's CipherFactory#CachedCipher
+    // (https://github.com/apache/cassandra/blob/trunk/src/java/org/apache/cassandra/security/CipherFactory.java)
+    private static final class CachedCipher {
+
+        private final int mode;
+        private final SecretKeySpec secretKeySpec;
+        private final Cipher cipher;
+
+        private CachedCipher(final int mode, final SecretKeySpec secretKeySpec, final Cipher cipher) {
+            this.mode = mode;
+            this.secretKeySpec = secretKeySpec;
+            this.cipher = cipher;
+        }
+
+        private boolean isCipherValid(final byte[] iv, final int cipherMode, final SecretKeySpec keySpec) {
+            return this.mode == cipherMode
+                    && this.secretKeySpec.equals(keySpec)
+                    && this.cipher.getAlgorithm().equals(ALGO_WITH_PADDING)
+                    && Arrays.equals(this.cipher.getIV(), iv);
+        }
+    }
+
+    private static final class CipherInitializationFailureException extends RuntimeException {
+
+        CipherInitializationFailureException(final Throwable cause) {
+            super(cause);
+        }
+    }
+
+    // Modified version of Cassandra's CipherFactory#buildCipher
+    // (https://github.com/apache/cassandra/blob/trunk/src/java/org/apache/cassandra/security/CipherFactory.java)
+    private Cipher buildCipher(final byte[] iv, final int cipherMode) {
         try {
-            this.cipher = Cipher.getInstance(ALGO_WITH_PADDING);
-        } catch (NoSuchAlgorithmException | NoSuchPaddingException e) {
-            LOGGER.error("Can not created instance of cipher with algorithm" + ALGO_WITH_PADDING);
+            CachedCipher cachedCipher = CIPHER_THREAD_LOCAL.get();
+
+            // Rigorous checks to make sure we've absolutely got the correct instance (with correct alg/key/iv/...)
+            if (cachedCipher != null && cachedCipher.isCipherValid(iv, cipherMode, this.secretKeySpec)) {
+                return cachedCipher.cipher;
+            }
+
+            Cipher cipher = Cipher.getInstance(ALGO_WITH_PADDING);
+            cipher.init(cipherMode, this.secretKeySpec, new IvParameterSpec(iv));
+            CIPHER_THREAD_LOCAL.set(new CachedCipher(cipherMode, this.secretKeySpec, cipher));
+            return cipher;
+        } catch (NoSuchAlgorithmException | NoSuchPaddingException | InvalidAlgorithmParameterException
+                | InvalidKeyException e) {
```

Comment:
It's better to catch a Throwable here and wrap anything that could go wrong with CipherInitializationFailureException